Contributions/Using_Linux_in_a_Safe_System.md

```diff
@@ -0,0 +1,348 @@
+
+* **Watchdogs / Safety Islands / Other Exception Contexts**
+
+  Continuing from the previous point, there are three typical design approaches,
+  for implementing the monitoring.
+  
+  *  **Watchdog**
+
+     A watchdog is a well known (simple) method for imposing mandatory behaviour on
+     a system; the basic idea is to implement a "dead-man button": if a component
+     is not producing a predefined stimuli, periodically, the watchdog will
+     inderogably trigger a predefined event.
+     Typically a hardware watchdog is a peripheral that enforces one of the
+     following policies:
+     *  NMI: the watchdog generates a Not Maskable Interrupt, which forces a core
+        to execute a predefined interrupt handler. The handler runs with its own
+        sanitsied context, and can be assumed to be in a sane/safe/secure state.
+        However, it retains also the access to the system state.
+
+        One could rely on the NMI interrupt service routine to take over and
+        initiate a reset, or power down, or whatever is required.
+     *  reset: it usually consists in the electrical state of the system being
+        taken to its default values, paired with software re-initialisation.
+     *  power down: power is cut off, at least from the components associated
+        to safety requirements.
+
+     This is fairly boilerplate information, but it is worth mentioning that
+     safety requirements apply not only to software but to hardware as well.
+     The Watchdog IP must be safety qualified, to be relied on.
+
+     However, for the watchdog to support safety claims, the following must
+     be proven:
+     *  the initialisation of the watchdog can be considered sufficiently safe
+        (same or better safety level than the required one)
+     *  it must be impossible for a single failure in the system to make it
+        ineffective; this case covers:
+        *  somehow a failure either disables or relaxes either the time
+           constraint of watchdog or action-upon-expiring
+        *  somehow a failure triggers one or more pings to the watchdog,
+           when instead it should expire
+
+     Note: it is safe to allow for a more restrictive re-configuring of the
+     watchdog, because even if it happened accidentally, it would only speed
+     up the expiration of the predefined timing period.
+
+     Continuing with the healthy approach of avoiding the introduction of
+     unnecessary dependency for the safety claims, the frugal designer will
+     recognise that the last thing one wants to do is to rely on Linux to
+     configure the watchdog, not only because of all the unnecessary safety
+     requirements placed on Linux, but also because such an approach would
+     leave a risk of having the system in an undefined state, potentially
+     indefinitely, should the linux boot fail before configuring successfully
+     the watchdog.
+     
```

Comment:
The above sentence could be misleading IMO.
In fact it could be perfectly fine to have an initial config of the WTD done in firmware and looking after the boot phase, while Linux can be used to re-configure the wtd in a second phase...